pkg/kubectl/cmd/apply.go

```diff
@@ -274,8 +279,39 @@ func (p *patcher) patch(current runtime.Object, modified []byte, source, namespa
 		if getErr != nil {
 			return nil, getErr
 		}
-		patchBytes, err = p.patchSimple(current, modified, source, namespace, name)
+		patchBytes, err = p.patchSimple(current, modified, source, namespace, name, resourceType)
 	}
 
 	return patchBytes, err
 }
+
+// checkIllegalChanges returns error when illegal updates are made, right now only
+// pod template labels updates and selector updates are disallowed.
+// TODO: do this in other kubectl update commands or move this to server-side
+func checkIllegalChanges(patch []byte, name, resourceType string) error {
+	diff := make(map[string]interface{})
+	if err := json.Unmarshal(patch, &diff); err != nil {
+		return fmt.Errorf("failed to unmarshal patch when determining if the change is legal: %v", err)
+	}
+	templateLabelsDiff := getNestedField(diff, "spec", "template", "metadata", "labels")
```

Comment:
At minimum move this to cmdutil.Factory and pass in the info.Mapping for the object.